crates/oxc_linter/src/rules/react/no_direct_mutation_state.rs

```diff
@@ -0,0 +1,460 @@
+    ///    },
+    ///    render: function() {
+    ///      return <div>Hello {this.state.name}</div>;
+    ///    }
+    ///  });
+    ///
+    ///  class Hello extends React.Component {
+    ///    constructor(props) {
+    ///      super(props)
+    ///
+    ///      doSomethingAsync(() => {
+    ///        this.state = 'bad';
+    ///      });
+    ///    }
+    ///  }
+    ///
+    ///  // success
+    ///  var Hello = createReactClass({
+    ///    componentDidMount: function() {
+    ///      this.setState({
+    ///        name: this.props.name.toUpperCase();
+    ///      });
+    ///    },
+    ///    render: function() {
+    ///      return <div>Hello {this.state.name}</div>;
+    ///    }
+    ///  });
+    ///
+    ///  class Hello extends React.Component {
+    ///    constructor(props) {
+    ///      super(props)
+    ///
+    ///      this.state = {
+    ///        foo: 'bar',
+    ///      }
+    ///    }
+    ///  }
+    /// ```
+    NoDirectMutationState,
+    correctness
+);
+
+// check current node is this.state.xx
+fn is_state_member_expression(expression: &StaticMemberExpression<'_>) -> bool {
+    if let Expression::ThisExpression(_) = &expression.object {
+        return expression.property.name == "state";
+    }
+
+    false
+}
+
+// get the top iterator
+// example: this.state.a.b.c.d => this.state
+fn get_outer_member_expression<'a, 'b>(
+    assignment: &'b SimpleAssignmentTarget<'a>,
+) -> Option<&'b StaticMemberExpression<'a>> {
+    if let SimpleAssignmentTarget::MemberAssignmentTarget(member_expr) = assignment {
+        match &member_expr.0 {
+            MemberExpression::StaticMemberExpression(expr) => {
+                let mut node = expr;
+                loop {
+                    if node.object.is_null() {
+                        return Some(node);
+                    }
+
+                    if let Some(object) = get_static_member_expression_obj(&node.object) {
+                        if !object.property.name.is_empty() {
+                            node = object;
+
+                            continue;
+                        }
+                    }
+
+                    return Some(node);
+                }
+            }
+            MemberExpression::PrivateFieldExpression(_)
+            | MemberExpression::ComputedMemberExpression(_) => {}
+        }
+    }
+
+    None
+}
+
+// Because node.object is of type &Expression<'_>
+// We need a function to get static_member_expression
+fn get_static_member_expression_obj<'a, 'b>(
+    expression: &'b Expression<'a>,
+) -> Option<&'b StaticMemberExpression<'a>> {
+    match expression {
+        Expression::MemberExpression(member_expr) => match &member_expr.0 {
+            MemberExpression::StaticMemberExpression(expr) => Some(expr),
+            _ => None,
+        },
+        _ => None,
+    }
+}
+
+fn should_ignore_component<'a, 'b>(node: &'b AstNode<'a>, ctx: &'b LintContext<'a>) -> bool {
+    let mut is_constructor: i8 = 0;
```

Comment:
any reason why you are using an `i8` instead of a bool?

Reply:
Hmm - using bool is a better solution, and I'll change that right away.